### PR TITLE
Auto-generate windows installer

### PR DESCRIPTION
**Pull request by BenWiederhake**
tl;dr: Windows installer.  Some questions at the bottom.

Since this is a major change (20 files changed, 390 insertions, 690 deletions), I think it's a better idea to first ask for more feedback.

What this does is simple to say: `./mkwindows.sh` generates a fresh, redistributable Windows installer from the current workdir.

Potential downsides:
- It's not as well-tested as Eion's hands.
- The third-party libraries about factor ten bigger, each.  Horrible.
- The build takes painstakingly long, around 5 minutes, as it needs to build quite a few things twice.
- It has to fumble with tgl's internal build system *by hand*.  But Eion had to do that already each time, so this is still faster and more robust.
- Building is Debian-specific, or at least probably difficult to do on Ubuntu.  I set up a virgin Debian VM and built everything from scratch to make sure that my instructions in the README are complete.  So I guess it's still more portable than our previous "Eion-only" build.
- Cross-compiling libwebp sucks.  But trying to use their `.lib` and `.dll`s does not work: some strange incompatibilities between crossbuilding-MingW and MSVC, which they used to compile their stuff.  And sure as hell I won't disable sticker support on Windows.

<3 Eion, I just want to take the pressure off you.  Your input is very valued, and you shouldn't need to waste your time building stuff on Windows.

So, open questions:
- We really should provide dev-builds of the installer.  Where do we put them?  @EionRobb, could we abuse your server space for that?
- Is there anything I could have missed?
- Can someone come up with a better idea than cross-compiling webp manually that actually works?

### CODE REVIEW

**Comment by BenWiederhake** [comment]
Plus of course:
- Does anyone care enough to figure out how to build it on Ubuntu, and write a Travis script for it?

**Comment by EionRobb** [comment]
> We really should provide dev-builds of the installer. Where do we put them? @EionRobb, could we abuse your server space for that?

Happy to host it

> Can someone come up with a better idea than cross-compiling webp manually that actually works?

Use an existing build of libwebp that's provided by mingw32 builds, eg opensuse's builds of it?

**Comment by EionRobb** [comment]
There's a lot of windows-specific changes in Makefile.in.  Would it be better to have a separate Makefile for win32-specific stuff that includes the base Makefile instead?

File: README.md [hunk 1]
```diff
@@ -257,6 +260,26 @@ and `.dsc` files, do this:
 
 Note that we no longer actively try to get it into the Debian repository.
 
+Building the Windows Installer
+---------------------------
+
+You will need a Debian-ish operating system, and the following packages:
+
+```
+apt-get install build-essentials gcc-mingw-w64-i686 libgcrypt-mingw-w64-dev \
```
Comment (EionRobb) [inline README.md]: Should this be suggesting 32bit deps instead of 64bit? (or can you not specify the w64 bit and let apt pick the most appropriate one)

File: mkwindows.sh [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+# Reconfigure
+echo "===== 4: Configure for cross-compilation"
+# Sadly, the library paths must be resolved *now* already.
+# AND, they must be absolute.  Sigh.
+export WIN32_GTK_DEV_DIR=`realpath win32/gtk_2_0-2.16`
+export WIN32_PIDGIN_DIR=`realpath win32/pidgin-2.12.0`
+export WIN32_WEBP_DIR="${WEBP_DIR_FULL}"
+./configure -q --host x86_64-linux-gnu --target i686-w64-mingw32 \
+    --includedir="/usr/share/mingw-w64/include" \
```
Comment (EionRobb) [inline mkwindows.sh]: can you use a pkg-config variable to find these paths instead?

**Comment by BenWiederhake** [comment]
> Use an existing build of libwebp that's provided by mingw32 builds, eg opensuse's builds of it?

Yeah, but I'm not exactly sure whether that really is okay license-wise, how to get the file during build time (downloading it could be seen as evil, including it in the repo is bloat).  I would also need all the headers with any configuration, if there is any (not sure).

Hmm.  "Download during build" sounds like the least-bad option to me.
What do you suggest?

> There's a lot of windows-specific changes in Makefile.in. Would it be better to have a separate Makefile for win32-specific stuff that includes the base Makefile instead?

Oh!  I didn't even think of that.  I'll do that, or the other way around.

File: mkwindows.sh (hunk 2 above)
Comment (BenWiederhake) [inline mkwindows.sh]: The first path?  No, since it's a build directory, but `configure` really dislikes relative paths.

The second one, given to `--includedir`?  Yeah, probably, I'll look into how well that works.  Thanks!

File: README.md (hunk 1 above)
Comment (BenWiederhake) [inline README.md]: For whatever reason I assumed that building 32-bit-only would be more reliable.

Should I create two installers, then?  One for 32-bits, one for 64?  After all, it's only double the time to create the installers, and double the space to store them, but not double the effort.  And useful for all those who have arrived in the 21st century and don't use 32-bit hardware for windows anymore :D

File: mkwindows.sh [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+cd ..
+
+# Pretend we're building up the preliminaries for auto/
+echo "===== 5: Compile tgl, pre-'generate' files"
+# Need to override LOCAL_LDFLAGS to remove '-rshared'.
+echo "    In case this fails mysteriously with some 'file stubs-32.h not found':"
+echo "    You're probably missing libc6-dev-i386 or some other dev files."
+make -C tgl -j4 LOCAL_LDFLAGS="-lz -lgcrypt -ggdb" bin/generate
+
+# Surprise!  Don't actually call bin/generate to create the autogen files,
```
Comment (EionRobb) [inline mkwindows.sh]: Why don't you want to generate?

File: telegram-purple.nsi [hunk 4]
```diff
@@ -92,23 +92,23 @@ Section "MainSection" SEC01
...
 	
 	SetOutPath "$PidginDir"
-	File "${WIN32_DEV_TOP}\libgpg-error-1.12-2\bin\libgpg-error-0.dll"
-	File "${WIN32_DEV_TOP}\libgcrypt-1.6.3\bin\libgcrypt-20.dll"
-	File "${WIN32_DEV_TOP}\libwebp-0.4.3-1\bin\libwebp-5.dll"
-	File "${WIN32_DEV_TOP}\mingw\bin\libgcc_s_sjlj-1.dll"
-	
+	File "${WIN32_DEV_TOP}\libgpg-error-0.dll"
+	File "${WIN32_DEV_TOP}\libgcrypt-20.dll"
+	File "${WIN32_DEV_TOP}\libwebp.dll"
```
Comment (EionRobb) [inline telegram-purple.nsi]: Shouldn't this be libwebp-5.dll?

File: mkwindows.sh (hunk 3 above)
Comment (BenWiederhake) [inline mkwindows.sh]: Because calling a Windows binary on Linux is quite a hassle.  And it doesn't work, as far as you have told me.

File: tgp-msg.c [hunk 5]
```diff
@@ -193,7 +193,16 @@ static char *format_document (const char *path, const char *filename, const char
   gchar *pth = g_markup_escape_text (path, -1);
   gchar *fle = g_markup_escape_text (filename, -1);
   gchar *mme = g_markup_escape_text (mime, -1);
-  gchar *fsize = g_format_size (size);
+  gchar *fsize =
+#if GLIB_CHECK_VERSION(2,30,0)
+    g_format_size (size)
+#elif GLIB_CHECK_VERSION(2,16,0)
+    /* Necessary as we compile on windows with glibc 2.24.2 */
```
Comment (EionRobb) [inline tgp-msg.c]: glib != glibc

Your version check is for 2.16 but the comment is 2.14?

File: telegram-purple.nsi (hunk 4 above)
Comment (BenWiederhake) [inline telegram-purple.nsi]: Good catch, but nope, I build libweb-6.dll, and since it's custom-built I thought I'd leave the 6 away.

Thinking about it again, I *really* should add the 6 back in.

This review is going awesome, thank you :D

File: tgp-msg.c (hunk 5 above)
Comment (BenWiederhake) [inline tgp-msg.c]: > glib != glibc

Oh shit.  Now I need to go through it all again, but I think the comment is wrong here.
Thanks for pointing it out.

> Your version check is for 2.16 but the comment is 2.14?

The comment is for version 2.24, since it's the version at which … you know what, I shouldn't have to explain this.  If I have to justify it *here*, then my code sucks.

I'm ashamed of not having written a comment.  I'll try again soon.  Spoiler: there's a good reason for the version mismatch.

File: README.md (hunk 1 above)
Comment (EionRobb) [inline README.md]: Sorry I was unclear, the question was asking whether it should be using Debian packages that don't have "w64" in the package name so that you can build on a non-64bit Linux machine, but now that I think about it, the packages have the same name on an i386 Debian too

File: mkwindows.sh (hunk 2 above)
Comment (EionRobb) [inline mkwindows.sh]: Yeah, the `--includedir` was the path I was referring to :)

File: mkwindows.sh (hunk 3 above)
Comment (EionRobb) [inline mkwindows.sh]: You're building generate for Linux (or more specifically, the host machine) though right?  You can then run it from linux/host to generate for the Windows build

Comment (BenWiederhake) [inline mkwindows.sh]: So instead of moving around the autogen files, it should move around `bin/generate`?  Hmm.  Actually, that seems much nicer, and gets rid of those ugly `touch` hacks a few lines down.  Thanks!

**Comment by EionRobb** [comment]
> Hmm. "Download during build" sounds like the least-bad option to me.

You're already doing that for the glib/gtk zip files from the BuildingWinPidgin steps, eg WIN32_PIDGIN_BIN_URL

> I'm not exactly sure whether that really is okay license-wise

You just need to be able to provide source when asked for GPL stuff, but libwebp is BSD so the same rules don't apply

**Comment by BenWiederhake** [comment]
I need an issue tracker within this issue.

- [ ] Use pre-built libwebp from … where?
- [ ] Find trustworthy, updating priviider of mingw builds of libwebp.
- [ ] Build both 32-bit and 64-bit
- [X] MinGW includedir: at least make it a Magic Constant
- [ ] Move win-specific `make` targets to own file
- [X] Instead of moving around the `auto` files, it should move around `bin/{generate,tl-parser}`
- [X] libwebp version-in-filename
- [ ] Check glibc vs glib
- [ ] Document new `#if GLIB_VERSION()` usage

Needed*

File: tgp-msg.c (hunk 5 above)
Comment (BenWiederhake) [inline tgp-msg.c]: > `/* Necessary as we compile on windows with glibc 2.24.2 */`

Huh, it already says that.
So was it all just a mis-read of `2.24` versus `2.14`?

Comment (EionRobb) [inline tgp-msg.c]: I probably bumped 1 instead of 2 when reviewing, but even still confusing comment between 2.24 vs 2.16

And the glibc thing still stands :)

**Comment by vk496** [comment]
Hello. 

I'm also trying to adapt the software for cross compile (in my case, `tg`). 

https://github.com/vk496/tg/tree/endian

Currently, I managed to compile `tl-parse` as native executable during cross compile. But I don't understand how `(Makefile.tgl)[https://github.com/vk496/tg/blob/endian/Makefile.tgl]` works and how use Makefiles rules to generate only `generate` binary and not for the other common rules. 

Would be nice make work native Linux cross compile. 

Salu2

**Comment by vk496** [comment]
Good news. I ported cross compiling to `tg` client.

[https://github.com/vk496/tg/tree/endian](https://github.com/vk496/tg/tree/endian)

Salu2

**Comment by BenWiederhake** [comment]
I'll try again in a different pull.  This is too messy.
